File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/recipe/GenerationsCoreRecipeSerializers.java
package generations.gg.generations.core.generationscore.common.world.recipe;

import dev.architectury.registry.registries.DeferredRegister;
import dev.architectury.registry.registries.RegistrySupplier;
import generations.gg.generations.core.generationscore.common.GenerationsCore;
import net.minecraft.core.registries.Registries;
import net.minecraft.world.item.crafting.Recipe;
import net.minecraft.world.item.crafting.RecipeSerializer;
import net.minecraft.world.item.crafting.ShapelessRecipe;

import java.util.function.Supplier;

public class GenerationsCoreRecipeSerializers {
	public static DeferredRegister<RecipeSerializer<?>> RECIPE_SERIALIZERS = DeferredRegister.create(GenerationsCore.MOD_ID, Registries.RECIPE_SERIALIZER);
	public static RegistrySupplier<RecipeSerializer<ShapedRksRecipe>> SHAPED_RKS = register("shaped_rks", ShapedRksRecipe.Serializer::new);
	public static RegistrySupplier<RecipeSerializer<ShapelessRksRecipe>> SHAPELESS_RKS = register("shapeless_rks", ShapelessRksRecipe.Serializer::new);

	public static void init() {
		RECIPE_SERIALIZERS.register();
	}

	public static <T extends Recipe<?>> RegistrySupplier<RecipeSerializer<T>> register(String name, Supplier<RecipeSerializer<T>> supplier) {
		return RECIPE_SERIALIZERS.register(name, supplier);
	}

}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/recipe/GenerationsCoreRecipeSerializers.kt
package generations.gg.generations.core.generationscore.common.world.recipe

import com.mojang.serialization.MapCodec
import generations.gg.generations.core.generationscore.common.util.PlatformRegistry
import net.minecraft.core.Holder
import net.minecraft.core.Registry
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.core.registries.Registries
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.resources.ResourceKey
import net.minecraft.world.item.crafting.Recipe
import net.minecraft.world.item.crafting.RecipeSerializer

object GenerationsCoreRecipeSerializers: PlatformRegistry<RecipeSerializer<*>>() {
    override val registry: Registry<RecipeSerializer<*>> = BuiltInRegistries.RECIPE_SERIALIZER
    override val resourceKey: ResourceKey<Registry<RecipeSerializer<*>>> = Registries.RECIPE_SERIALIZER

    var SHAPED_RKS = register("shaped_rks", ShapedRksRecipe.CODEC, ShapedRksRecipe.STREAM_CODEC)
    var SHAPELESS_RKS = register("shapeless_rks", ShapelessRksRecipe.CODEC, ShapelessRksRecipe.STREAM_CODEC)

    fun <T : Recipe<*>> register(
        name: String,
        codec: MapCodec<T>,
        streamCodec: StreamCodec<RegistryFriendlyByteBuf, T>
    ): Holder<RecipeSerializer<*>> = create(name, {
        object : RecipeSerializer<T> {
            override fun codec(): MapCodec<T> {
                return codec
            }

            override fun streamCodec(): StreamCodec<RegistryFriendlyByteBuf, T> {
                return streamCodec
            }
        }
    })
}
